### ISSUE TITLE
ClosedSendChannelException in PingPong.kt

### ISSUE THREAD

**Issue by frececroka** [ktorio/ktor#1346]
**Ktor Version and Engine Used (client or server and name)**

1.2.4, client, ktor-client-cio, ktor-websockets

**Describe the bug**

The `CoroutineScope.ponger` function sometimes sends a pong to a closed channel. This happens when the `outgoing` channel is already closed while the ponger is still active. This throws an exception:

```
Exception in thread "Thread-33259" kotlinx.coroutines.channels.ClosedSendChannelException: Channel was closed
	at kotlinx.coroutines.channels.Closed.getSendException(AbstractChannel.kt:1048)
	at kotlinx.coroutines.channels.AbstractSendChannel.offer(AbstractChannel.kt:166)
	at kotlinx.coroutines.channels.AbstractSendChannel.send(AbstractChannel.kt:146)
	at io.ktor.http.cio.websocket.PingPongKt$ponger$1.invokeSuspend(PingPong.kt:42)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:241)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:835)
```

**To Reproduce**

A simple server that exposes one WebSocket endpoint:

```kotlin
import io.ktor.application.install
import io.ktor.routing.routing
import io.ktor.server.engine.embeddedServer
import io.ktor.server.netty.Netty
import io.ktor.websocket.WebSockets
import io.ktor.websocket.webSocket
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.channels.consumeEach
import kotlinx.coroutines.runBlocking
import java.time.Duration

@ExperimentalCoroutinesApi
fun main() { runBlocking {
    val server = embeddedServer(Netty, 8080) {
        install(WebSockets) {
            pingPeriod = Duration.ofMillis(200)
        }
        routing {
            webSocket("/ws") {
                incoming.consumeEach {}
            }
        }
    }
    server.start(true)
} }
```

A client that hammers the server with requests:

```kotlin
import io.ktor.client.HttpClient
import io.ktor.client.features.websocket.WebSockets
import io.ktor.client.features.websocket.ws
import io.ktor.http.HttpMethod
import io.ktor.util.KtorExperimentalAPI
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlin.random.Random.Default.nextInt

@KtorExperimentalAPI
fun main() { runBlocking {
    repeat(100) { launch {
        while (true) {
            val client = HttpClient { install(WebSockets) }
            client.use {
                it.ws(method = HttpMethod.Get, host = "localhost", port = 8080, path = "/ws") {
                    delay(nextInt(980, 1020).toLong())
                }
            }
        }
    } }
} }
```

Ktor dependencies:
```kotlin
implementation("io.ktor:ktor-client-cio:1.2.4")
implementation("io.ktor:ktor-server-netty:1.2.4")
implementation("io.ktor:ktor-websockets:1.2.4")
```

**Expected behavior**

I believe sending a pong to a closed channel should not produce an exception.

**Comment by cy6erGn0m** [comment]
fixed in PR #1332

**Comment by frececroka** [comment]
Thanks, is the PR also catching exceptions thrown by the `HttpClient`? It looks like it concerns only server code.

**Comment by cy6erGn0m** [comment]
ping/pong and close sequence handling in both server and client

**Comment by cy6erGn0m** [comment]
You still can have `ClosedSendChannelException` in your code at `send` invocation. Should be fixed in PR #1348

**Comment by frececroka** [comment]
Just tested it with current master (`1.3.0-beta-1`) and it works perfectly. Thanks, very nice :+1:.
